docs/latex/Inhalt/03_Inhalt/02_ddd.tex

```diff
@@ -0,0 +1,35 @@
+
+\section{Entities \& Value Objects}
+
+\begin{figure}[ht!]
+    \includegraphics[width=0.98\columnwidth]{../diagrams/entity_uml.pdf}
+    \caption{Klassendiagramm der Domäne}
+    \label{fig:class-diag-domain}
+\end{figure}
+
+\autoref{fig:class-diag-domain} zeigt das Klassendiagramm der Domäne. Im Package recipe existiert eine Entity: das Recipe. Jedes Recipe besitzt durch eine UUID eine eigene Identität. Außerdem können Rezepte in seltenen Fällen auch nachträglich abgewandelt werden und besitzen daher grundsätzlich veränderliche Werte. Bei der Ingredient hingegen handelt es sich um ein Value Object, da diese keine eigene Identität besitzt. Existieren zwei Ingredients mit dem gleichen Food und der gleichen Quantity, so sind diese gleich. Außerdem ist ihr Zustand unveränderlich, so wird etwa bei einer Änderung in einem Rezept die betreffende Ingredient gelöscht und eine neue erstellt. Ebenso verhält es sich mit der Quantity: Sind Amount und Unit einer Quantity gleich, so sind auch die Quantities gleich und auch ihr Zustand ist unveränderlich. Zuletzt enthält das Package noch die Unit. Auch hier handelt es sich um ein Value Object, da auch die Gleichheit von Units lediglich von ihren Werten, also von ihren Namen abhängt. Ebenso besitzt eine Unit keinen eigenen Lebenszyklus und ihr Zustand ist unveränderlich.
+
+Auch im Package profile existiert eine Entity: das Profile. Jedes Profile besitzt durch seinen Namen eine eigene Identität, da auch zwei Profiles mit gleichem vorhandenen Food und gleichen Opinions zu verschiedenen Nutzern gehören können und daher unterscheidbar sein müssen. Die auch die Unterklassen der abstrakten Klasse Opinion, welche die möglichen Opinions eines Nutzers über ein Food abbilden, repräsentieren Value Objects, da sie weder eine eigene Identität noch veränderliche Werte besitzen. Da sie außerdem keine Attribute besitzen, wird ihre Gleichheit nur über ihren Typ bestimmt. Die Verbindung zwischen Food, Profile und Opinion wird durch OpinionAbout hergestellt. Es handelt sich auch hier um eine Value Object, da auch eine OpinionAbout weder einen eigenen Lebenszyklus noch eine Identität besitzt, da sie durch ihre Werte, also das Food und die zugehörige Opinion, definiert wird. Solle ein Nutzer seine Meinung zu einem Food ändern, so wird die bestehende gelöscht und eine neue erzeugt.
```

Comment:
Klassennamen, packagenamen bitte in \code

Reply:
Es wird nicht klar erklärt, warum in der Domäne eine OpinionAbout ist, das wirkt mir wie etwas technisches. Wollen wir die nicht einfach wegwerfen und durch eine Map von Food auf Opinion ersetzen (zu jedem Food hat ein Profile eine Opinion, das klingt in der Dömane sinnvoller, finde ich)

Reply:
Fliegt raus. Ich wusste auch nicht wirklich, wie ich das als Domain Code verkaufen soll.